urlfcc351a1ba_kuntapavankumar_Z_tgz-pJ8-edu_concordia_dpis_ReplicaTestJ8/src/edu/concordia/dpis/stationserver/StationServerImpl.java

```diff
@@ -0,0 +1,741 @@
+								buf.append(",");
+								buf.append(cRecord.getFirstName());
+								buf.append(",");
+								buf.append(cRecord.getLastName());
+								buf.append(",");
+								buf.append(cRecord.getDescription());
+								buf.append(",");
+								buf.append(cRecord.getStatus());
+
+							} else {
+								MissingRecord mRecord = (MissingRecord) record;
+								buf.append(",");
+								buf.append(mRecord.getFirstName());
+								buf.append(",");
+								buf.append(mRecord.getLastName());
+								buf.append(",");
+								buf.append(mRecord.getLastKnownAddress());
+								buf.append(",");
+								buf.append(mRecord.getLastSeenDate());
+								buf.append(",");
+								buf.append(mRecord.getLastSeenPlace());
+								buf.append(",");
+								buf.append(mRecord.getStatus());
+							}
+							log.debug("Requested record transfer..."
+									+ buf.toString());
+							// send over tcp
+							result = client.transfer(buf.toString(),
+									StationType
+											.valueOf(remoteStationServerName));
+							ItemtoRemove = record;
+							break;
+						}
+					}// while
+					if (result) {
+						// if successful remove the record from this list
+						boolean isRemoved;
+						synchronized (ItemtoRemove) {
+							isRemoved = recList.remove(ItemtoRemove);
+						}
+						log.debug("Transferred successfully");
+						return isRemoved;
+					} else {
+						log.error("Could not transfer record");
+					}
+				}
+			}
+		} else {
+			log.error("Officer with badgeId " + badgeId
+					+ " does not belong to this station");
+		}
+		return false;
+	}
+
+	// return ths number of records in this stations
+	private long countRecords() {
+		long no_of_records = 0;
+		for (Character ch : records.keySet()) {
+			no_of_records += this.records.get(ch).size();
+		}
+		return no_of_records;
+	}
+
+	// a TCP Client program to invoke the TCP server on the other side
+	class TCPClient {
+		private boolean transfer(String record, StationType station) {
+			Socket s = null;
+
+			String[] hostAndPort = extractHostAndPortNumber((String) tcpProperties
+					.get(station.getStationCode().toString()));
+			int serverPort = Integer.parseInt(hostAndPort[1]);
+			try {
+				s = new Socket(hostAndPort[0], serverPort);
+				// input stream
+				DataInputStream in = new DataInputStream(s.getInputStream());
+				// output stream
+				DataOutputStream out = new DataOutputStream(s.getOutputStream());
+
+				out.writeUTF(record);
+				boolean result = Boolean.parseBoolean(in.readUTF());
+				return result;
+			} catch (IOException e) {
+				e.printStackTrace();
+				log.error(e.getMessage());
+			}
+			return false;
+		}
+	}
+
+	// A UDPClient to invoke the UDPServer
+	class UDPClient {
+
+		private String ping(String remoteHost, int serverPort)
+				throws FileNotFoundException, IOException {
+			String response = "";
+			Properties properties = new Properties();
+			properties.load(new FileInputStream("udp.properties"));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource created by the following constructor: *FileInputStream*. Execution paths that do not contain closure statements were detected. Ensure the object returned by *FileInputStream()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).